### PR TITLE
Set texture

### PR DESCRIPTION
**Pull request by patriciogonzalezvivo**
Dynamically change textures using an ```url```/```image```/```canvas```/```data``` by providing one of this on the ```scene.config.textures```.  Ex:

```
scene.config.textures.orbits.url = "http://..../image.png";
scene.updateConfig({ rebuild: true });

// or

var canvas = new document.createElement("canvas");
...
scene.config.textures.orbits.element = canvas;
scene.updateConfig({ rebuild: true });

// or

var image = new document.createElement("image");
...
scene.config.textures.orbits.element = image;
scene.updateConfig({ rebuild: true });

// or

var raw_data = { data: [], width: 500, height: 500 }
...
scene.config.textures.orbits.data = raw_data;
scene.updateConfig({ rebuild: true });
```

### CODE REVIEW

File: src/gl/texture.js [hunk 1]
```diff
@@ -16,19 +16,20 @@ export default class Texture {
             this.valid = true;
         }
         this.bind();
-        this.image = null;      // an Image object/element that is the source for this texture
-        this.canvas = null;     // a Canvas object/element that is the source for this texture
+
+        this.name = name;
+        this.source = null;
```
Comment (bcamper) [inline src/gl/texture.js]: Let's use `null` instead of strings like "none", that's more JS-like.

File: src/scene.js [hunk 2]
```diff
@@ -1000,6 +1000,28 @@ export default class Scene {
         }
     }
 
+    // setTexture (style_name, uniform_name, element) {
```
Comment (bcamper) [inline src/scene.js]: revert these changes

File: src/scene_loader.js [hunk 3]
```diff
@@ -86,7 +86,10 @@ export default SceneLoader = {
 
                     // Texture by URL, expand relative to scene file
                     if (typeof tex === 'string' && !config.textures[tex]) {
-                        style.texture = Utils.addBaseURL(tex, path);
+                        let texture_name = '__' + style_name;
+                        config.textures[texture_name] = { url: tex };
+                        style.texture = texture_name; // point style to location of texture
+                        // style.texture = Utils.addBaseURL(tex, path);
```
Comment (bcamper) [inline src/scene_loader.js]: we can take out these commented lines

File: src/scene_loader.js [hunk 4]
```diff
@@ -115,11 +122,14 @@ export default SceneLoader = {
...
-                                style.material[prop].texture = Utils.addBaseURL(tex, path);
+                                let texture_name = '__' + style_name + '_material_' + prop;
+                                config.textures[texture_name] = { url: tex };
+                                style.material[prop].texture = texture_name; // point style to location of texture
+                                // style.material[prop].texture = Utils.addBaseURL(tex, path);
                             }
                             // Texture by object, move it to the global scene texture set and give it a default name
                             else if (typeof tex === 'object') {
-                                let texture_name = '__' + style_name;
+                                let texture_name = '__' + style_name + '_material_' + prop;
```
Comment (bcamper) [inline src/scene_loader.js]: Note: this fixes an existing bug in master.

File: src/scene_loader.js [hunk 5]
```diff
@@ -99,10 +102,14 @@ export default SceneLoader = {
                 // If style has texture uniforms, expand texture URLs relative to scene file
                 if (style.shaders && style.shaders.uniforms) {
                     for (let {type, value, key, uniforms} of GLSL.parseUniforms(style.shaders.uniforms)) {
-                        if (type === 'sampler2D' && !config.textures[value]) {
-                            uniforms[key] = Utils.addBaseURL(value, path);
+                        if (type === 'sampler2D' && typeof value === 'string' && !config.textures[value]) {
+                            let texture_name = '__' + style_name + '_uniform_' + key;
```
Comment (bcamper) [inline src/scene_loader.js]: Note: this fixes an existing bug in master.

File: src/gl/texture.js [hunk 6]
```diff
@@ -299,6 +316,7 @@ Texture.createFromObject = function (gl, textures) {
 Texture.changed = function (name, config) {
     if (Texture.textures[name]) { // cached texture
         // compare definitions
+        // TODO: fix for canvas/image/data
```
Comment (bcamper) [inline src/gl/texture.js]: `canvas`/`image` elements won't be stringified, which means that differences in them won't be detected. We probably need to go ahead and write (or use an existing) deep comparison method.

File: src/gl/texture.js (hunk 1 above)
Comment (patriciogonzalezvivo) [inline src/gl/texture.js]: :+1:

File: src/gl/texture.js (hunk 6 above)
Comment (patriciogonzalezvivo) [inline src/gl/texture.js]: Is not easier to add boolean flag for changes?

**Comment by bcamper** [comment]
Can you please add a description of changes, and usage examples?

**Comment by bcamper** [comment]
Please also remember to not commit the `dist/` files in the future, we should remove them from this PR, but even adding them in a single commit adds them to history forever :/

**Comment by patriciogonzalezvivo** [comment]
Should I delete and create again this branch?

**Comment by bcamper** [comment]
I don't think so, we can rebase before merging.

**Comment by patriciogonzalezvivo** [comment]
sorry, I don't remember when I include them

**Comment by bcamper** [comment]
We don't commit them except when a new release is cut'm

**Comment by bcamper** [comment]
Now that we have this working, can we combine `canvas` and `image` into a single `element` parameter for textures? We can detect based on type, and other aspects are similar (both can be provided directly as WebGL textures).

**Comment by patriciogonzalezvivo** [comment]
Done! Also, that lint error on text.js is annoying. Any idea how to solve it?

**Comment by bcamper** [comment]
Re: lint, just remove `let texture = ` and let the `new Texture` call be on its own. I don't really like that syntax wise, but it's a result of how the rest of the system works. Maybe I can replace with a static instantiation method that would be clearer (`Texture.create`).

**Comment by patriciogonzalezvivo** [comment]
:+1:

**Comment by patriciogonzalezvivo** [comment]
should I merge master on it? or you want to do it before merging this brunch into master?

**Comment by bcamper** [comment]
No please wait, I will review more later, thanks.

**Comment by bcamper** [comment]
@patriciogonzalezvivo with https://github.com/tangrams/tangram/commit/4172dc5534f89740ee3ba09d106536c54c3ee59d, the `element` can be set to a CSS selector string. 

So you can setup HTML like:

`<canvas class='awesome-texture'>`

And then reference it as a texture with:

```
textures:
   awesome-texture:
      element: .awesome-texture
```

Or you could reference an element by id, but the use of `#` in CSS selector means you need to be careful to quote the YAML so it's not interpreted as a comment :) For example:

`<canvas id='texture-id'>`

```
textures:
   awesome-texture:
      element: '#texture-id'  # without quote, the # will be interpreted as a YAML comment!
```

**Comment by patriciogonzalezvivo** [comment]
:+1: nice!

**Comment by patriciogonzalezvivo** [comment]
yay!
